### PR TITLE
Fix ships being wrong unit type

### PR DESCRIPTION
**Pull request by walterroach**
Fixes not being able to target ships #259 

This overwrites the inherited init and add_unit methods in DDGroupGenerator from GroupGenerator, which assumes Vehicle type (DDGroupGenerator should create type Ship).

Other objects inheriting from GroupGenerator may also suffer this problem, but I haven't yet investigated further.

### CODE REVIEW

**Comment by walterroach** [comment]
Actually, there are still cases where ships are not given correct type during start generation.  It looks like we need to make a new ship generator class that applies the same changes I made to DDGroupGenerator, and have all the other naval generator classes inherit the add_unit and init methods.

File: gen/fleet/dd_group.py [hunk 1]
```diff
@@ -1,20 +1,41 @@
 import random
 
 from gen.sam.group_generator import GroupGenerator
+from dcs import unitgroup
+from dcs.point import PointAction
+from dcs.unit import Ship
 from dcs.ships import *
 
 
 class DDGroupGenerator(GroupGenerator):
```
Comment (DanAlbert) [inline gen/fleet/dd_group.py]: Is DD the base for all ships, or is that why this is still WIP? I know we also spawn FFGs and CGs.

Probably a good idea to make `GroupGenerator` define `def create_group(...) -> Group` with no implementation and leave that to `ShipGroupGenerator` and `VehicleGroupGenerator` (or `GroundUnitGenerator`?) just to avoid confusion.

File: gen/fleet/dd_group.py [hunk 2]
```diff
@@ -1,20 +1,41 @@
...
+        wp.ETA_locked = True
         self.ddtype = ddtype
 
     def generate(self):
         self.add_unit(self.ddtype, "DD1", self.position.x + 500, self.position.y + 900, self.heading)
         self.add_unit(self.ddtype, "DD2", self.position.x + 500, self.position.y - 900, self.heading)
         self.get_generated_group().points[0].speed = 20
+    
+    def add_unit(self, unit_type, name, pos_x, pos_y, heading):
+        nn = "cgroup|" + str(self.go.cp_id) + '|' + str(self.go.group_id) + '|' + str(self.go.group_identifier) + "|" + name
```
Comment (DanAlbert) [inline gen/fleet/dd_group.py]: Unrelated, but do we need to name these so they can contribute to Skynet's EWR? @davidp57

File: gen/fleet/dd_group.py (hunk 1 above)
Comment (walterroach) [reply gen/fleet/dd_group.py]: Some ship groups inherit from DDGroupGenerator, others inherit from GroupGenerator.

I'm unclear on if this is intentional or not.  79bbef08 moves all those that inherited from GroupGenerator over to a new ShipGroupGenerator with the same changes as the original PR applied to them.

This appears to be working properly, but if anyone knows the reason for DDGroupGenerator they should inspect that correct behavior is occurring.

Currently, create_group is done during init on all three of these related classes (DDGroupGenerator, GroupGenerator, and ShipGroupGenerator), but I agree moving it to a method create_group would make it both easier to read and simpler to extend.

**Comment by walterroach** [comment]
Last commit is just comment removal and addition of type hinting to make it easier to pick apart further later.  

This should be good to go.

File: gen/fleet/dd_group.py (hunk 1 above)
Comment (walterroach) [reply gen/fleet/dd_group.py]: I think building a create_group method can kill a few birds at once here, related to problems with SkyNet and SEAD/DEAD missions being sometimes broken.  

The easy fix for these issues, is during create_group(), we also check if it's a SAM that should participate in SkyNet, then update the corresponding TheaterGroundObject.name with the same prefix that is applied to the pydcs VehicleGroup object name.

This ~~will~~ could break anything that uses the TheaterGroundObject.name value _before_ the generator is called though, so I'm not sure this is the best option.

Comment (walterroach) [reply gen/fleet/dd_group.py]: #298 #66

File: gen/fleet/dd_group.py (hunk 2 above)
Comment (Khopa) [reply gen/fleet/dd_group.py]: We could add an additional "|" separator at the end and whatever skynet needs after that.

File: gen/fleet/dd_group.py (hunk 1 above)
Comment (Khopa) [reply gen/fleet/dd_group.py]: > Some ship groups inherit from DDGroupGenerator, others inherit from GroupGenerator.
> I'm unclear on if this is intentional or not.

I think this was intended to avoid rewriting the same code for ships group that use the same formation with a single unit.
DDGroupGenerator was just a more specialized GroupGenerator with a unit type given. No big deal there.
The others kind of ship groups generators (ChineseNavyGroupGenerator, RussianNavyGroupGenerator for instance aren't used much because it resulted in a lot of ships being added)

But the whole thing is messy. So feel free to refactor ;)

Comment (Khopa) [reply gen/fleet/dd_group.py]: I'm unsure what kind of data skynet needs. Whether the unit has a radar or not to contribute to the EWR networks, and that's it ? Or is there more to it ?

Comment (DanAlbert) [reply gen/fleet/dd_group.py]: > This will could break anything that uses the TheaterGroundObject.name value before the generator is called though, so I'm not sure this is the best option.

Wouldn't be my first choice. That bug sounds really easy to write. Probably nothing in `TheaterGroundObject` aside from `is_dead` should be mutable.

Instead, why don't we restructure that so that the group generation happens _before_ creating the TGO? They currently happen at the same time, just in the other order, and I don't think there's actually any need for that. Tough to do before I submit the refactor that I've been working on. Haven't had much time for the last few days but I'll polish that up soon and we can follow up on this after if that works for you.

> I'm unsure what kind of data skynet needs. Whether the unit has a radar or not to contribute to the EWR networks, and that's it ? Or is there more to it ?

It needs units to have predictable name prefixes that indicate what type of thing they are (we can alternatively give it a list of units, but the prefix is easier IMO and also plays nicely with MOOSE if plugins want to use that). So units need a prefix of `RED SAM ` or `BLUE EWR ` etc.

Comment (walterroach) [reply gen/fleet/dd_group.py]: It appears like there is two options I can see for adding SAM sites to SkyNet:
1. Use a uniform prefix on the DCS group name, then define the prefix(s) in the Lua setup script ( the method that's currently in use), then define the prefix(s)
2. Explicitly add each group by specifying the full DCS group name 

To exercise option 2 in Liberation, we'd need to dynamically generate a bunch of Lua in Liberation (undesirable I think).

The problem we have with the current implementation is that group names are shared between pydcs objects and liberation objects, and modifying one isn't assured to update state in the other.

**Review by DanAlbert — APPROVED** [review]

File: gen/fleet/dd_group.py (hunk 1 above)
Comment (davidp57) [reply gen/fleet/dd_group.py]: Both methods are good, imo.
Method 1 is already in place, but as you said it needs the names to be predefined.
Method 2 should be cleaner, but needs some engine work (I don't feel adequate for that).
I already inject lua tables based on the units generated by liberation (not only the SAMs), and it works well (although it may not be very elegant, I used the Python objects after they're generated, it'd be cleaner to have some kind of "database" object that everyone calls when spawning something, to construct a complete list with all the needed detail).

So, to sum up, if you feel like refactoring the generation engine to fill up a nice database with all the units details (group and unit name, function, position, radio freqs, the whole nine yards), it'd be the best option imho.
I'd follow up by modifying the plugins and using this data.
